### ISSUE TITLE
Incorrect results on 1.0

### ISSUE THREAD

**Issue by inducer** [pocl/pocl#594]
I've just migrated our infrastructure over to Pocl 1.0, and I'm seeing a bunch of odd test failures:

https://gitlab.tiker.net/inducer/pytential/pipelines/9793

(Mainly, incorrectly computed results.) Now, my sense is that this has only started occurring once after I've enabled testing with the conda-packaged version of pocl. Normally, most of the tests run with a machine-specific pocl built on the host itself, but the conda tests use a "distro build". Now I'm wondering whether something could go wrong if those two pocls start sharing a cache...?

I'd be grateful for any hints.

In the meantime, I'll force the conda builds onto a separate cache dir and see if that improves matters. I'll report back.

**Comment by inducer** [comment]
@mattwala raised a valid point in https://gitlab.tiker.net/inducer/pytential/issues/75#note_20563: We also share PyOpenCL caches between those, and so one may be fed binaries coming from the other. Maybe the platform name should change based on the type of build. (This would allow pyopencl to not mix and match binaries.)

**Comment by inducer** [comment]
Update: It doesn't look like keeping the caches apart has helped. So there's potentially something else fishy.

https://gitlab.tiker.net/inducer/pytential/issues/75

**Comment by inducer** [comment]
I'll close this since I'm not sure I have a viable reproducer. But the good news I spread around after the RC1 release announcement may have been premature.

**Comment by franz** [comment]
@inducer  binary format changed between 0.14 and 1.0, and it's mutually incompatible. WRT "Maybe the platform name should change based on the type of build." - platform *version* contains build-specific information, we may add it there.

On the machine that has failing tests, you run them in parallel yes ?

**Comment by franz** [comment]
BTW - LLVM IR format has changed in the past, and it may change again. Since pocl binaries contain LLVM IR, you cannot reliably pass pocl binaries between different pocl major versions, even if we froze pocl's binary format. The only safe way would be to use SPIR-V but that's not yet implemented.

**Comment by inducer** [comment]
> platform version contains build-specific information, we may add it there.

That'd be fine. PyOpenCL uses [every bit of identifying information](https://github.com/inducer/pyopencl/blob/40d40b89ab79c6e1d68b1b63a2d4bfa489ef55f5/pyopencl/cache.py#L235-L240) it can get its hands on.

> binary format changed between 0.14 and 1.0

As long as one of `platform.vendor, platform.name, platform.version,
device.vendor, device.name, device.version, device.driver_version` changed, that's OK from PyOpenCL's end. It won't serve up stale binaries.

> BTW - LLVM IR format has changed in the past, and it may change again. Since pocl binaries contain LLVM IR, you cannot reliably pass pocl binaries between different pocl major versions, even if we froze pocl's binary format.

Is there some tuple of values (maybe the ones I give above) that, by being identical, indicates that pocl binaries are safe to share?

As a side note, I tried separating out the caches for each CI run (so no concurrent access), and that did not improve matters. Downgrading to pocl 0.13 (with the same code on the Python side) helped, though.

**Comment by franz** [comment]
> Is there some tuple of values (maybe the ones I give above) that, by being identical, indicates that pocl binaries are safe to share?

At least pocl version + llvm version + CPU type, but probably there are more. In fact there's already a mechanism in pocl to deal with this, but it likely needs some tuning.

> As a side note, I tried separating out the caches for each CI run (so no concurrent access), and that did not improve matters

Not surprising. You said "Mainly, incorrectly computed results" - the cache is possible, but unlikely culprit here. If there's a cache failure, it usually leads to abort, so you'd have no results at all. It's possible there's some mismatch between CPU capabilities / kernel library / binaries, but i'm not sure how it can give worse results than pocl 0.13... is the test that's failing public ? is it failing also on non-distro builds ?

**Comment by inducer** [comment]
> At least pocl version + llvm version + CPU type, but probably there are more. In fact there's already a mechanism in pocl to deal with this, but it likely needs some tuning.

Split into a separate issue: #596

> is the test that's failing public ?

Yes, it is. See https://github.com/inducer/pytential and install info linked there. https://gitlab.tiker.net/inducer/pytential/issues/75 for context.

> is it failing also on non-distro builds ?

Yes, it is.

The failures currently only seem to happen deep some of our most complicated application code, only sporadically, and typically only when run by our CIs. I have not observed these failures in person on POCL 1.0. Given the complexity of that code base, I suspect we'll realistically have to do some more legwork on our end to reproduce this before you can meaningfully jump in. I do admit that I'm a little at a loss for how to track this down.

**Comment by pjaaskel** [comment]
Have you tried if this happens with the unthreaded 'basic' pocl device driver? Sporadic indeterministic issues can be also race conditions in the task scheduling code of the pthread driver.

**Comment by inducer** [comment]
@pjaaskel Thanks for the hint! That allowed me to get somewhere with this.

As a test case, I use a batch of tests from [pytential](https://github.com/inducer/pytential) revision 5a96eda with this patch applied:

```diff
diff --git a/test/test_layer_pot_identity.py b/test/test_layer_pot_identity.py
index fe00134..a7e8b68 100644
--- a/test/test_layer_pot_identity.py
+++ b/test/test_layer_pot_identity.py
@@ -250,7 +250,7 @@ class DynamicTestCase(object):
     tc
     for geom in [
         StarfishGeometry(),
-        SphereGeometry(),
+        #SphereGeometry(),
         ]
     for tc in [
         DynamicTestCase(geom, GreenExpr(), 0),
@@ -259,8 +259,8 @@ class DynamicTestCase(object):
         DynamicTestCase(geom, GradGreenExpr(), 1.2),
         DynamicTestCase(geom, ZeroCalderonExpr(), 0),
 
-        DynamicTestCase(geom, GreenExpr(), 0, fmm_backend="fmmlib"),
-        DynamicTestCase(geom, GreenExpr(), 1.2, fmm_backend="fmmlib"),
+        #DynamicTestCase(geom, GreenExpr(), 0, fmm_backend="fmmlib"),
+        #DynamicTestCase(geom, GreenExpr(), 1.2, fmm_backend="fmmlib"),
         ]])
 def test_identity_convergence(ctx_getter,  case, visualize=False):
     logging.basicConfig(level=logging.INFO)
```

(I imagine the patch is non-essential, it just disables some of the more long-running/complex tests.)

I run the following command:

```shell
while true; do PYOPENCL_TEST=port python -m pytest -p no:warnings test_layer_pot_identity.py; done
```

Crucially, I was only able to observe failures if the machine is heavily loaded. To generate a dummy load, I use this script:

```python
#! /usr/bin/python3
from multiprocessing import Pool, TimeoutError
import time
import os

def f(x):
    for i in range(10**6):
        pass

if __name__ == '__main__':
    pool = Pool()

    while True:
        for i in pool.imap_unordered(f, range(10**6)):
            pass
        print("completed a round")
```

I am not sure it matters, but the machine I tested on has a large-ish number of cores. It is a dual-socket E5-2650 v4 machine. (so, 24 cores with HT disabled) I'm on pocl c8e0dba0 compiled from source with LLVM 5.0.0 from Debian packages, built using [this script](https://gist.github.com/inducer/ff35d807fe2f58227908ee179058a45f).

With the command above, perhaps one in five runs would result in a test failure.

By comparison, this command:
```
while true; do POCL_DEVICES="basic" PYOPENCL_TEST=port python -m pytest -p no:warnings test_layer_pot_identity.py; done
```
would not yield any failures through roughly 20 runs of the batch of tests. (I'll keep this running to confirm and update.)

**Comment by inducer** [comment]
As an update, the tests with the basic driver continued to run for about an hour (at about one cycle per minute) without any failures observed.

**Comment by inducer** [comment]
> Crucially, I was only able to observe failures if the machine is heavily loaded.

This also explains why these failures were seen from our CIs (which spawn a number of concurrent jobs) but not in person (where I would typically pick a less-busy machine).

**Comment by pjaaskel** [comment]
OK. If you can help more, it might be worthwhile to review the pthread driver code and focus on the mutual exclusion parts.

**Comment by inducer** [comment]
So I started by doing the thing which requires zero skill, which is put pocl with a demo kernel through [helgrind](http://valgrind.org/docs/manual/hg-manual.html):

It found two things that looked a bit fishy:

```
==15092== Possible data race during read of size 1 at 0xCE3E9FAD0 by thread #36                                                                                                                                                                                                 
==15092== Locks held: none                                                                      
==15092==    at 0x4C2EB40: my_memcmp (hg_intercepts.c:211)                                                                                          
==15092==    by 0x4C2EE9F: mutex_destroy_WRK (hg_intercepts.c:850)                                                                                                                  
==15092==    by 0x476674BC: POclReleaseKernel (in /opt/pocl-release_1_0/lib/libpocl.so.2.0.0)                                                                      
==15092==    by 0x476988D3: pthread_scheduler_get_work (in /opt/pocl-release_1_0/lib/libpocl.so.2.0.0)                                                                                                                                        
==15092==    by 0x47698BC2: pocl_pthread_driver_thread (in /opt/pocl-release_1_0/lib/libpocl.so.2.0.0)                                                                                                                                       
==15092==    by 0x4C31D36: mythread_wrapper (hg_intercepts.c:389)                                                                                                                          
==15092==    by 0x4E45518: start_thread (pthread_create.c:456)                                       
==15092==                                                                                                                                                
==15092== This conflicts with a previous write of size 4 by thread #1   
==15092== Locks held: none                                                                                                                                                             
==15092==    at 0x4E4949C: __pthread_mutex_unlock_usercnt (pthread_mutex_unlock.c:49)
==15092==    by 0x4C2F696: mutex_unlock_WRK (hg_intercepts.c:1097)        
==15092==    by 0x47667392: POclReleaseKernel (in /opt/pocl-release_1_0/lib/libpocl.so.2.0.0)
==15092==    by 0x46530A43: _call_func<int (*)(_cl_kernel*), 0, _cl_kernel* const&> (function.h:38)                                                                                                                                          
==15092==    by 0x46530A43: call_tuple<int (*&)(_cl_kernel*), std::tuple<_cl_kernel* const&> > (function.h:49)
==15092==    by 0x46530A43: call<__CLArgGetter, int (*)(_cl_kernel*)> (function.h:108)
==15092==    by 0x46530A43: clcall<int (*)(_cl_kernel*)> (error.h:211)      
==15092==    by 0x46530A43: call_guarded_cleanup<_cl_kernel*, _cl_kernel* const&> (error.h:281)   
==15092==    by 0x46530A43: kernel::~kernel() (kernel.cpp:19)                                    
==15092==    by 0x46530CD8: kernel::~kernel() (kernel.cpp:20)                                                                                                                              
==15092==    by 0x464EF4BB: _cffi_f_clobj__delete(_object*, _object*) (pyopencl._cffi.cpp:1358)
==15092==    by 0x4C4E9C: _PyCFunction_FastCallKeywords (in /usr/bin/python3.6)                                                                          
==15092==    by 0x54F973: ??? (in /usr/bin/python3.6)                                              
==15092==  Address 0xce3e9fad0 is 16 bytes inside a block of size 152 alloc'd                                    
==15092==    at 0x4C2E015: calloc (vg_replace_malloc.c:711)                                                                                                                            
==15092==    by 0x47666B02: POclCreateKernel (in /opt/pocl-release_1_0/lib/libpocl.so.2.0.0) 
==15092==    by 0x46534D52: _call_func<_cl_kernel* (*)(_cl_program*, char const*, int*), 0, 1, 2, _cl_program* const&, char const*&, int*> (function.h:38)
==15092==    by 0x46534D52: call_tuple<_cl_kernel* (*&)(_cl_program*, char const*, int*), std::tuple<_cl_program* const&, char const*&, int*> > (function.h:49)
==15092==    by 0x46534D52: call<__CLArgGetter, _cl_kernel* (*)(_cl_program*, char const*, int*)> (function.h:108)
==15092==    by 0x46534D52: clcall<_cl_kernel* (*)(_cl_program*, char const*, int*)> (error.h:211)     
==15092==    by 0x46534D52: call_guarded<_cl_kernel*, _cl_program*, char const*, int*, const program*&, char const*&> (error.h:257)
==15092==    by 0x46534D52: operator() (kernel.cpp:99)                         
==15092==    by 0x46534D52: c_handle_error<create_kernel(clbase**, clobj_t, char const*)::<lambda()> > (error.h:296)
==15092==    by 0x46534D52: create_kernel (kernel.cpp:101)                                             
==15092==    by 0x464ED989: _cffi_f_create_kernel(_object*, _object*) (pyopencl._cffi.cpp:2311)
==15092==    by 0x4C4DFC: _PyCFunction_FastCallKeywords (in /usr/bin/python3.6)                
==15092==    by 0x54F973: ??? (in /usr/bin/python3.6)                                       
==15092==    by 0x55405E: _PyEval_EvalFrameDefault (in /usr/bin/python3.6)       
==15092==    by 0x54EA77: ??? (in /usr/bin/python3.6)                                         
==15092==    by 0x558881: _PyFunction_FastCallDict (in /usr/bin/python3.6)                                                                                    
==15092==    by 0x459FB0: _PyObject_Call_Prepend (in /usr/bin/python3.6)                                                                                           
==15092==    by 0x459A3D: PyObject_Call (in /usr/bin/python3.6)                                                          
==15092==    by 0x4E0F1A: ??? (in /usr/bin/python3.6)                                                    
==15092==  Block was alloc'd by thread #1           
```

and

```
==15092== Possible data race during read of size 1 at 0xC5AF99850 by thread #1
==15092== Locks held: none
==15092==    at 0x4C2EB40: my_memcmp (hg_intercepts.c:211)
==15092==    by 0x4C2EE9F: mutex_destroy_WRK (hg_intercepts.c:850)
==15092==    by 0x47670EF6: POclReleaseEvent (in /opt/pocl-release_1_0/lib/libpocl.so.2.0.0)
==15092==    by 0x4650D128: _call_func<int (*)(_cl_event*), 0, _cl_event* const&> (function.h:38)
==15092==    by 0x4650D128: call_tuple<int (*&)(_cl_event*), std::tuple<_cl_event* const&> > (function.h:49)
==15092==    by 0x4650D128: call<__CLArgGetter, int (*)(_cl_event*)> (function.h:108)
==15092==    by 0x4650D128: clcall<int (*)(_cl_event*)> (error.h:211)
==15092==    by 0x4650D128: call_guarded_cleanup<_cl_event*, _cl_event* const&> (error.h:281)
==15092==    by 0x4650D128: event::~event() (event.cpp:109)
==15092==    by 0x4650EB96: ~nanny_event (event.h:73)
==15092==    by 0x4650EB96: nanny_event::~nanny_event() (event.h:73)
==15092==    by 0x464EF4BB: _cffi_f_clobj__delete(_object*, _object*) (pyopencl._cffi.cpp:1358)
==15092==    by 0x4C4E9C: _PyCFunction_FastCallKeywords (in /usr/bin/python3.6)
==15092==    by 0x54F973: ??? (in /usr/bin/python3.6)
==15092==    by 0x55405E: _PyEval_EvalFrameDefault (in /usr/bin/python3.6)
==15092==    by 0x54EA77: ??? (in /usr/bin/python3.6)
==15092==    by 0x558881: _PyFunction_FastCallDict (in /usr/bin/python3.6)
==15092==    by 0x459FB0: _PyObject_Call_Prepend (in /usr/bin/python3.6)
==15092==
==15092== This conflicts with a previous write of size 4 by thread #38
==15092== Locks held: none
==15092==    at 0x4E4949C: __pthread_mutex_unlock_usercnt (pthread_mutex_unlock.c:49)
==15092==    by 0x4C2F696: mutex_unlock_WRK (hg_intercepts.c:1097)
==15092==    by 0x47670E77: POclReleaseEvent (in /opt/pocl-release_1_0/lib/libpocl.so.2.0.0)
==15092==    by 0x47694ADA: pocl_exec_command (in /opt/pocl-release_1_0/lib/libpocl.so.2.0.0)
==15092==    by 0x47698E77: pocl_pthread_driver_thread (in /opt/pocl-release_1_0/lib/libpocl.so.2.0.0)
==15092==    by 0x4C31D36: mythread_wrapper (hg_intercepts.c:389)
==15092==    by 0x4E45518: start_thread (pthread_create.c:456)
==15092==  Address 0xc5af99850 is 16 bytes inside a block of size 200 alloc'd
==15092==    at 0x4C2E015: calloc (vg_replace_malloc.c:711)
==15092==    by 0x4767C08F: pocl_mem_manager_new_event (in /opt/pocl-release_1_0/lib/libpocl.so.2.0.0)
==15092==    by 0x4767663B: pocl_create_event (in /opt/pocl-release_1_0/lib/libpocl.so.2.0.0)
==15092==    by 0x4767681B: pocl_create_command (in /opt/pocl-release_1_0/lib/libpocl.so.2.0.0)
==15092==    by 0x4765FE3B: POclEnqueueReadBuffer (in /opt/pocl-release_1_0/lib/libpocl.so.2.0.0)
==15092==    by 0x465212F4: _call_func<int (*)(_cl_command_queue*, _cl_mem*, unsigned int, long unsigned int, long unsigned int, void*, unsigned int, _cl_event* const*, _cl_event**), 0, 1, 2, 3, 4, 5, 6, 7, 8, _cl_command_queue* const&, _cl_mem* const&, unsigned int&, lon
g unsigned int&, long unsigned int&, void*&, long unsigned int, _cl_event**, _cl_event**> (function.h:38)
==15092==    by 0x465212F4: call_tuple<int (*&)(_cl_command_queue*, _cl_mem*, unsigned int, long unsigned int, long unsigned int, void*, unsigned int, _cl_event* const*, _cl_event**), std::tuple<_cl_command_queue* const&, _cl_mem* const&, unsigned int&, long unsigned int&
, long unsigned int&, void*&, long unsigned int, _cl_event**, _cl_event**> > (function.h:49)
==15092==    by 0x465212F4: call<__CLArgGetter, int (*)(_cl_command_queue*, _cl_mem*, unsigned int, long unsigned int, long unsigned int, void*, unsigned int, _cl_event* const*, _cl_event**)> (function.h:108)
==15092==    by 0x465212F4: clcall<int (*)(_cl_command_queue*, _cl_mem*, unsigned int, long unsigned int, long unsigned int, void*, unsigned int, _cl_event* const*, _cl_event**)> (error.h:211)
==15092==    by 0x465212F4: call_guarded<command_queue*&, memory_object*&, bool, long unsigned int&, long unsigned int&, void*&, const pyopencl_buf<_cl_event*>&, _CLObjOutArg<nanny_event, void*>, _cl_command_queue*, _cl_mem*, unsigned int, long unsigned int, long unsigned
 int, void*, unsigned int, _cl_event* const*, _cl_event**> (error.h:243)
==15092==    by 0x465212F4: operator() (buffer.cpp:57)
==15092==    by 0x465212F4: retry_mem_error<enqueue_read_buffer(clbase**, clobj_t, clobj_t, void*, size_t, size_t, clbase* const*, uint32_t, int, void*)::<lambda()> > (error.h:319)
==15092==    by 0x465212F4: operator() (error.h:332)
==15092==    by 0x465212F4: c_handle_error<c_handle_retry_mem_error(Func&&) [with Func = enqueue_read_buffer(clbase**, clobj_t, clobj_t, void*, size_t, size_t, clbase* const*, uint32_t, int, void*)::<lambda()>]::<lambda()> > (error.h:296)
==15092==    by 0x465212F4: c_handle_retry_mem_error<enqueue_read_buffer(clbase**, clobj_t, clobj_t, void*, size_t, size_t, clbase* const*, uint32_t, int, void*)::<lambda()> > (error.h:332)
==15092==    by 0x465212F4: enqueue_read_buffer (buffer.cpp:58)
==15092==    by 0x464E9195: _cffi_f_enqueue_read_buffer(_object*, _object*) (pyopencl._cffi.cpp:4635)
==15092==    by 0x4C4DFC: _PyCFunction_FastCallKeywords (in /usr/bin/python3.6)
==15092==    by 0x54F973: ??? (in /usr/bin/python3.6)
==15092==    by 0x55405E: _PyEval_EvalFrameDefault (in /usr/bin/python3.6)
==15092==    by 0x54F570: ??? (in /usr/bin/python3.6)
==15092==    by 0x55059D: PyEval_EvalCodeEx (in /usr/bin/python3.6)
==15092==  Block was alloc'd by thread #1
```

Do these look incriminating to you?

**Comment by franz** [comment]
> Do these look incriminating to you?

Not really, there were a zillion similar errors in older pocl versions, but everything is possible with threads. Helgring is slow as hell, isn't it ? if i may suggest ThreadSanitizer instead, it likely has much more reasonable overhead. Compile pocl with ``-DENABLE_TSAN=1 -DENABLE_ICD=0`` ; compile PyOpenCL (or your C / C++ program) against lib/CL/libOpenCL.so from previous step, not ocl-icd;  then run your program with ``LD_PRELOAD=/path/to/libtsan.so`` usually it's ``/usr/lib/libtsan.so``. libtsan must be the first library loaded by ld, otherwise it doesn't work.

BTW is the machine where you get these failures a NUMA machine ?

**Comment by inducer** [comment]
> BTW is the machine where you get these failures a NUMA machine ?

It is. Why would it matter?

**Comment by pjaaskel** [comment]
Looks fishy indeed. It should give good hints where to review for a possible race to too early destruction, which might well cause indeterminism. Seeing similar errors in earlier pocl versions and not hitting the issue with this test case might be just pure luck.

**Comment by franz** [comment]
@inducer i tried to install pytential with ``python setup.py install``

but it ends up with an error

```
Searching for sumpy>=2013.1
Reading https://pypi.python.org/simple/sumpy/
No local packages or working download links found for sumpy>=2013.1
error: Could not find suitable distribution for Requirement.parse('sumpy>=2013.1')
```

**Comment by inducer** [comment]
Sorry! All our stuff is still sufficiently beta that we haven't really bothered with a release to the package index. These install instructions should work:

https://documen.tician.de/pytential/misc.html

Once that's all done, you can go to `YOUR_CONDA_ROOT/envs/inteq/etc/OpenCL/vendors` and configure it to point to your dev pocl. I'd be delighted to help you reproduce this, so if there's anything I can do, just let me know.

**Comment by franz** [comment]
BTW the "fishy" things found by helgrind are false alarm. Looking at last few lines in each backtrace, you'll notice both are in ``clReleaseSomething``, one in ``pthread_mutex_destroy`` the other in ``pthread_mutex_unlock``.

Our code looks like

```
  pthread_mutex_lock(obj);
  int new_refcount = DEC_REFCOUNT(obj);
  pthread_mutex_unlock(obj);
  if (new_refcount == 0)
    { ... pthread_mutex_destroy(obj->lock);  }
```

Since we're calling ``pthread_mutex_destroy`` on an unlocked object, it could *in theory* race with another ``pthread_mutex_lock`` of the same object (by some other API call). But because we're using refcounts, this could only happen if any API is called after refcount drops to zero. All of ``clRelease*`` calls are guarded against this by an ``assert()`` check in ``DEC_REFCOUNT`` (which i don't remember seeing fail, ever). I'll add another assert (into all API calls, not just clRelease*) to make sure none of the APIs are called with refcount==0, but i'm doubtful this is the actual issue with pytential.

**Comment by franz** [comment]
I've managed to reproduce the error with "test_exterior_stokes_2d", getting about 80% fail rate on a busy machine (actually it fails a lot on idle too, so this is likely not related to machine load..)

Unfortunately nothing popped up so far. I've added a ton of asserts, still nothing. ThreadSanitizer-built pocl prints a ton of data races, but after i fixed two in pocl, the remaining ones look like this:

```
  Write of size 4 at 0x7ffe42cca1a0 by main thread:
    #0 POclGetDeviceInfo /home/devel/0/pocl/lib/CL/clGetDeviceInfo.c:64 (libOpenCL.so+0x000000348b87)
    #1 device::get_info(unsigned int) const <null> (_cffi.abi3.so+0x00000003616e)

  Previous read of size 8 at 0x7ffe42cca1a0 by thread T19:
    #0 memcpy /build/gcc/src/gcc/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:596 (libtsan.so.0+0x00000002fda2)
    #1 mkl_blas_dgemm_omp_driver_v1 <null> (libmkl_intel_thread.so+0x0000002faa0b)
```

If anyone knows what is "libmkl_intel_thread.so" and how can it interfere with pocl, that might be helpful.

I'll run with AddressSanitizer-built pocl, if that fails i guess it's down to git bisect..

**Comment by franz** [comment]
ASAN-enabled pocl reports no errors & no warnings.

**Comment by inducer** [comment]
> If anyone knows what is "libmkl_intel_thread.so" and how can it interfere with pocl, that might be helpful.

Intel MKL is an implementation of [BLAS](https://en.wikipedia.org/wiki/Basic_Linear_Algebra_Subprograms) (among other things), which is used by numpy for, in this case, matrix multiplication ("[dgemm](https://www.math.utah.edu/software/lapack/lapack-blas/dgemm.html)").

FWIW, I'm seeing these failures on a numpy build that doesn't use MKL, so it's unlikely that what we're seeing is specific to MKL.

In addition, if real, that's a really odd race. I would expect that dgemm would mostly handle floating point data, while GetDeviceInfo mostly returns integers...

**Comment by franz** [comment]
> so it's unlikely that what we're seeing is specific to MKL.

@inducer yes that seems correct, i've rerun the test with TSan and POCL_DEVICES=basic, same data races but the results is not an error, so it's unlikely the problem is any of those data races. Since both TSan and ASan come up empty, it's more likely some logic error, rather than data race.

**Comment by pjaaskel** [comment]
It's suspicious, though, if we cannot reproduce this with the 'basic' driver.

**Comment by franz** [comment]
@pjaaskel  well, looking at API trace of a test run, it seems to create & destroy a huge number of command queues. The pthread driver only guarantees order within a single command queue, not among multiple command queues. So this may be some interaction bug in pocl or even in the test itself. Could be why it always works with basic driver...

**Comment by pjaaskel** [comment]
@franz Yep. If the app erroneusly assumes any implicy ordering between execution of commands submitted to different CQs it can manifest as a task execution race that is hidden by 'basic'.

**Comment by franz** [comment]
Just to check the theory, i've added a ``printf()`` into the driver to print the number of commands in flight. When it's passing, the max is 2 commands, with about 100 printfs in the entire log; when failing it has a max of 5 commands with nearly 11000 printfs. I'm guessing the linux kernel scheduler schedules the driver threads less frequently, thus their queue grows bigger.

So this may be some missing dependency bug in pocl, or a missing dependency bug in the test. Unfortunately the way opencl is used (and where the command queues are created) in the python test is non-obvious to me.

**Comment by pjaaskel** [comment]
(Like I wrote in the IRC:) A possible easy way to add a sanity check for the basic multiple-CQ cross-queue race condition is to check for each written (or possibly written) buffer argument if it's read in another CQ and there is no explicit event synchronization ordering the commands. The app might assume there is implicit data dependence based resolution also across CQs which there isn't (and is not mandated by the specs).

**Comment by inducer** [comment]
Yikes. @franz's comment regarding the many command queues reminded me of something, and I've got a mounting body of evidence that this is our fault after all. I'm very sorry! I'll keep you updated, and if there's a way for me to buy you pizza or beer or some goodies of your choice for your group meeting, I would love to do so. (Do let me know!)

In the meantime, thank you very much for your help with this.

**Comment by inducer** [comment]
It looks like [this](https://gitlab.tiker.net/inducer/sumpy/merge_requests/67) was at fault, partially enabled by the lack of [this](https://gitlab.tiker.net/inducer/loopy/merge_requests/205).

**Comment by franz** [comment]
@inducer Well, at least it's resolved.. I've also fixed a few other potential races, so wasn't entirely wasted time.

**Comment by pjaaskel** [comment]
This just hints that pocl is reaching the point where the reported bugs are more and more likely to be client side issues, which is good news to me ;-)
